app/models/workflow.rb

```diff
@@ -23,7 +24,19 @@ def external_signal?
     template&.signal_setting.present?
   end
 
+  def deletable?
+    requests.any? { |request| !request.finished? } ? false : true
+  end
+
   private
+  
+  def validate_deletable
+    throw :abort unless deletable?
+
+    requests.select { |request| request.finished? }.each do |request|
+      EventService.new(request).workflow_deleted
```

Comment:
I'm confused by a few things here:
1. The `deletable?` method is already checking for finished request and we would not get here if there are still pending requests.  So do we need the select loop at all?  (More in this loop in the next item)
2. I feel like the event should be a life-cycle event on the workflow object and I would not expect a separate event per-request it is linked to.  Meaning, I was expecting a single "workflow deleted" event message.
3. If someone needs to know that the workflow was removed from the completed request (not sure we have a use-case right now) I would expect that to be multiple events against the request object.

cc @mkanoor @eclarizio

Reply:
If a `workflow` is deletable, we only update its associated `requests`, nullify their `workflow_id`, not delete them. I will change to broadcast single message in this case.